### PR TITLE
[flutter_tools] return terminal to echo and line mode before exiting resident_runner

### LINKED ISSUE #105255: flutter tool is turning echo off in the terminal in regular usage
After running `flutter run` and the app is quitting normally I'm not getting local echo on my terminal (I can't see what I'm typing).

`reset` will fix the issue.

Output before the echo is turned off:
```
Flutter run key commands.
r Hot reload. 🔥🔥🔥
R Hot restart.
h List all available interactive commands.
d Detach (terminate "flutter run" but leave application running).
c Clear the screen
q Quit (terminate the application on the device).

💪 Running with sound null safety 💪

An Observatory debugger and profiler on Aaron’s iPhone is available at: http://127.0.0.1:58590/KS8qMQD_RrI=/
The Flutter DevTools debugger and profiler on Aaron’s iPhone is available at:
http://127.0.0.1:9102?uri=http://127.0.0.1:58590/KS8qMQD_RrI=/

Application finished.
aaclarke-macbookpro2:counter aaclarke$ 
```

I'm not sure if it happens 100% of the time but it is happening at least most of the time for me.

doctor:
```
[✓] Flutter (Channel unknown, 3.1.0-0.0.pre.1072, on macOS 12.4 21F79 darwin-x64, locale en-US)
    • Flutter version 3.1.0-0.0.pre.1072 at /Users/aaclarke/dev/flutter
    • Upstream repository unknown
    • Framework revision ff110fb958 (2 hours ago), 2022-06-02 13:11:28 -0700
    • Engine revision 00a64bf4f2
    • Dart version 2.18.0 (build 2.18.0-163.0.dev)
    • DevTools version 2.13.1
```

cc @christopherfujino

### PR DESCRIPTION
**Pull request by christopherfujino**
Fixes https://github.com/flutter/flutter/issues/105255

### CODE REVIEW

File: packages/flutter_tools/lib/src/commands/run.dart [hunk 1]
```diff
@@ -715,6 +715,10 @@ class RunCommand extends RunCommandBase {
         throwToolExit('Lost connection to device.');
       }
       rethrow;
+    } finally {
+      // However we exited from the runner, ensure the terminal has line mode
+      // and echo mode enabled before we return the user to the shell.
+      globals.terminal.singleCharMode = false;
```
Comment (zanderso) [inline packages/flutter_tools/lib/src/commands/run.dart:721]: Does the resident runner modify any other modes that should be reset?

**Review by zanderso — APPROVED** [review]
lgtm w/ question

File: packages/flutter_tools/lib/src/commands/run.dart (hunk 1 above)
Comment (christopherfujino) [reply packages/flutter_tools/lib/src/commands/run.dart:721]: So this is actually a convenience setter around two modes:

```
    if (value) {
      stdin.echoMode = false;
      stdin.lineMode = false;
    } else {
      stdin.lineMode = true;
      stdin.echoMode = true;
    }
```

https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/terminal.dart#L288

I'll take one more pass at the code to check for anything else suspicious, however.

**Comment by zanderso** [comment]
FYI @aam

**Pull request merged**